### PR TITLE
fix(signals): deduplicate collision clusters, label buckets

### LINKED ISSUE #440: Track public-surface truth projection and panel consistency regressions
## Summary

Recent PR and review evidence points at one recurring maintenance problem: internal truth is correct in one layer, then a public or maintainer-facing projection drifts.

Concrete examples:

- #412 hardened the public-surface publishing path after live smoke testing showed a PR comment-write 403 could abort other outputs.
- #419 / #420 showed `REQUIRED_INSTALLATION_PERMISSIONS` correctly required `pull_requests: write`, while the repo settings preview still told maintainers `pull_requests: read` and omitted the active missing permission.
- #417 / #418 showed contributor `dominantLabels` double-counted overlapping activity sources and could bias `labelFit` scoring.
- #424 / #425 showed the public PR panel counted duplicate risk with `Math.max(...)` while its details listed the deduped union.
- #416 and #426 show the public PR panel is now carrying high-value maintainer context, including high/critical queue state and large overlap clusters, so count/text consistency matters.

## Suggested work

- Add a small dev-facing checklist or test helper for source-of-truth projection: shared constant or canonical collection -> internal decision -> API/settings preview -> public bot text -> fail-on-revert coverage.
- Add golden consistency coverage for public PR panel sections where a headline count and a detail list are derived from the same conceptual set.
- Prefer reusing canonical permission constants and precomputed union/dedup collections instead of reconstructing equivalent logic in render layers.
- Make advisory-vs-blocking wording part of the assertions, not just the numeric values.

## Acceptance criteria

- Permission surfaces cannot again say `pull_requests: read` when the canonical app contract requires write for PR comment/label output.
- Public panel duplicate/overlap counts are tested against the exact collection rendered in details.
- At least one regression test fails if a future render path computes a count from a different source than the details section.
- Maintainer-facing remediation text names the real blocking source, not only the aggregate status.

### PR DESCRIPTION
**Pull request by dale053**
## Summary

- Fix `requiredInstallPermissions` in `settings-preview.ts` to derive read-only base permissions from `REQUIRED_INSTALLATION_PERMISSIONS` instead of hardcoded strings, so the preview can't silently drift from the canonical constant (regression from issue #419/#420 pattern).
- Fix label double-counting in `refreshContributorActivity`: `allPullRequests`, `mergedPullRequests`, and `openPullRequests` are overlapping views of the same PR set — deduplicate nodes by URL before extracting labels so a label on one PR isn't counted three times (regression from issue #417/#418 pattern).
- Fix `scopedOverlapCount` in the gate check path of `processors.ts`: replace `Math.max(prSpecificCount, preflight.collisions.length)` with the same Map-deduped union that `buildPublicPrIntelligenceComment` uses, so the readiness score matches the rendered panel details (regression from issue #424/#425 pattern).
- Add three regression tests (one per fix) to `backfill.test.ts`, `settings-preview.test.ts`, and `signals-coverage.test.ts`.

## Scope

- [x] This PR is focused and does not mix unrelated backend, UI, MCP, docs, dependency, and deploy changes.
- [x] This follows `CONTRIBUTING.md` and does not reintroduce GitHub Pages, VitePress, `site/`, or `CNAME`.
- [x] I linked an issue, or this is small enough that the summary explains why an issue is not needed.

## Validation

- [ ] `git diff --check`
- [ ] `npm run actionlint`
- [ ] `npm run typecheck`
- [ ] `npm run test:coverage` locally; global coverage stays at or above **97%** for lines, statements, functions, and branches (aim for **98%+** branch coverage locally so CI variance does not fail near the threshold)
- [ ] `npm run test:workers`
- [ ] `npm run build:mcp`
- [ ] `npm run test:mcp-pack`
- [ ] `npm run ui:openapi:check`
- [ ] `npm run ui:lint`
- [ ] `npm run ui:typecheck`
- [ ] `npm run ui:build`
- [ ] `npm audit --audit-level=moderate`
- [x] New or changed behavior has unit/integration tests for new branches, fallback paths, and sanitizer boundaries

If any required check was skipped, explain why:

- UI checks not applicable — no UI changes in this PR.

## Safety

- [x] No secrets, wallet details, hotkeys, coldkeys, user PATs, private keys, raw trust scores, private rankings, or private maintainer evidence are exposed.
- [x] Public GitHub text stays sanitized, low-noise, and does not imply compensation guarantees or optimization tactics.
- [ ] Auth, cookie, CORS, GitHub App, Cloudflare, or session changes include negative-path tests. *(not applicable)*
- [ ] API/OpenAPI/MCP behavior is updated and tested where needed. *(not applicable)*
- [ ] UI changes use live API data or real empty/error/loading states, not production mock/demo fallbacks. *(not applicable)*
- [ ] Visible UI changes include screenshots or a short recording. *(not applicable)*
- [ ] Public docs/changelogs are updated where needed; changelogs are only edited for release-prep PRs. *(not applicable)*

## Notes

- Closes #440.
- The `Math.max` → union change in `processors.ts` is a pure correctness fix; the readiness score total is unaffected when the two sets are disjoint (the common case), and is now correct when they overlap.
- The permission surface fix filters only `read`-valued entries from `REQUIRED_INSTALLATION_PERMISSIONS` as the base, keeping `issues: write` conditional on whether output is actually enabled — preserving the existing behavior while ensuring future additions to the constant are automatically reflected.

### CODE REVIEW

**Comment by ghost** [comment]
> [!NOTE]
> ## Gittensory Gate skipped
> PR closed before full evaluation. No late first comment was created.
>
> | Signal | Result | Evidence | Action |
> | --- | --- | --- | --- |
> | Gate result | ⚠️ Skipped | JSONbored/gittensory#446 is no longer open. | No action. |

---
Checked by [Gittensory](https://github.com/JSONbored/gittensory), a quiet PR intelligence layer for OSS maintainers.

File: package-lock.json [hunk 1]
```diff
@@ -1497,6 +1497,7 @@
       "cpu": [
         "ppc64"
       ],
+      "dev": true,
       "license": "MIT",
       "optional": true,
```
Comment (JSONbored) [inline package-lock.json:1503]: This lockfile churn appears unrelated to the PR: no dependency or package-manager input changed, but many optional platform entries gained/lost metadata. Please remove this file from the branch unless the lockfile update is intentional and explained.

File: test/unit/signals-coverage.test.ts [hunk 2]
```diff
@@ -1112,6 +1112,51 @@ describe("signal coverage edge cases", () => {
...
+    // The panel should show the cluster once (deduplicated union = 1), not twice (max gives 1 here too,
+    // but the key invariant is the comment's overlap count equals the union count).
+    const overlapMatch = comment.match(/(\d+) scoped overlap/);
+    const unionCount = [...new Map([...(prCluster ? [prCluster] : []), sharedCluster].map((c) => [c.id, c])).values()].length;
+    expect(overlapMatch ? Number(overlapMatch[1]) : 0).toBeLessThanOrEqual(unionCount);
+  });
+
   it("filters disabled linked-issue findings and uses fallback next steps when the panel is clean", () => {
     const directRepo = repo("owner/clean-panel");
     const currentPr = pr(directRepo.fullName, 50, "Fix documented bug", {
```
Comment (JSONbored) [inline test/unit/signals-coverage.test.ts:1296]: This does not fail on the production regression it describes. The changed code path is `processors.ts` computing `scopedOverlapCount` before calling `buildPublicReadinessScore`, but this test calls `buildPublicPrIntelligenceComment` directly and only asserts `<= unionCount`, so reverting the processor to `Math.max(...)` would still pass. Please test the extracted union helper or the processor/public-surface path directly.

**Review by JSONbored — CHANGES_REQUESTED** [review]
Thanks for picking this up. The direction matches #440, but I found two blockers before this should merge:

1. `package-lock.json` has broad optional dependency metadata churn even though the PR does not change dependencies. Please drop that file from the PR unless there is a concrete package-manager reason tied to this fix.
2. The scoped-overlap regression test does not exercise the production path changed in `processors.ts`, and the final assertion can pass even if the processor goes back to `Math.max(...)`. Please either extract the scoped-overlap union helper and test that directly, or add a focused processor-level/public-surface test that fails if `scopedOverlapCount` is computed from `Math.max(prSpecificCount, preflight.collisions.length)` again.

The label dedupe and permission-constant direction look reasonable once those two issues are fixed.

**Review by JSONbored — CHANGES_REQUESTED** [review]
@dale053 the direction is still useful, and the unrelated lockfile churn is gone from the current diff.

A few notes:
- The label dedupe and permission-constant changes are reasonable.
- The production code now computes the scoped-overlap union instead of `Math.max(...)`.
- The remaining substantive blocker is still regression coverage for that production path. The new test exercises rendered comment behavior and can pass without proving `maybePublishPrPublicSurface` feeds the deduped scoped-overlap count through the live webhook processor path.
- The PR title also still contains a literal ellipsis, so it does not meet the Conventional Commit title rule.

Required changes:
- Test the extracted scoped-overlap union helper directly, or add a focused processor/public-surface test that fails if the code returns to `Math.max(...)`.
- Retitle the PR to `fix(signals): deduplicate collision clusters and permission buckets`.

Validation expected:
- Rerun the focused backfill, settings-preview, and signals/public-surface tests plus standard validation.

**Review by JSONbored — CHANGES_REQUESTED** [review]
@dale053 the source diff is focused and the production-path overlap coverage is much better than the earlier version.

A few notes:
- I tried GitHub's update-branch rebase before reviewing because the PR is behind current main, and GitHub reported a rebase conflict.
- I am not going to resolve that conflict from the maintainer side.
- After a clean rebase, this is likely reviewable again without changing the product direction.

Required changes:
~~~text
- Rebase onto current main and resolve the conflict locally.
- Preserve the backfill label dedupe regression, scoped-overlap union coverage, and settings-preview permission regression coverage.
- Re-run the focused signals/backfill/settings-preview tests and full validation.
~~~

Validation expected:
~~~text
- git diff --check
- focused backfill/signals/settings-preview tests
- full validate
~~~

**Review by JSONbored — CHANGES_REQUESTED** [review]
@dale053 this now looks good to land.

A few notes:
- The unrelated lockfile churn is gone.
- The scoped-overlap union is now in the production queue path rather than only comment rendering.
- Regression coverage now exercises the dedupe behavior that was previously missing.

Fix failing CI.

**Comment by dale053** [comment]
Hi,@JSONbored!
Could you please review this PR?

**Review by JSONbored — APPROVED** [review]
@dale053 this is good to land.

A few notes:
- The label dedupe now handles overlapping PR buckets before deriving dominant labels.
- The scoped-overlap count now uses the same union logic in the production queue path and rendered panel path.
- The settings preview now derives read permissions from the canonical install-permission constant.
- The regression tests cover the three drift cases called out in #440.

No code changes requested.
